### PR TITLE
Updated with JS example for testing authorization

### PR DESCRIPTION
**Pull request by flinkgutt**

### CODE REVIEW

File: request-authentication.md [hunk 1]
```diff
@@ -6,13 +6,85 @@ pagination_next: null
...
 webhook secret.
 
+<Tabs
+    defaultValue="js" 
+    groupId="example-choice" 
+    values={[ {label: 'JavaScript', value: 'js'}, {label: '.Net C#', value: 'csharp'}, ]}>
+
+<TabItem value="js">
+
+To run: `node webhook.mjs`
```
Comment (maro-vm) [inline request-authentication.md]: It's not clear that webhook.mjs is the file below.
I vote for just removing the the 'To run' section, or alternatively add a header to clearly indicate this is webhook.mjs.

File: request-authentication.md [hunk 2]
```diff
@@ -6,13 +6,85 @@ pagination_next: null
...
+```js
+import crypto from "crypto";
+
+// The secret is found in the response body when you register a webhook. 
+// Note: The secret is unique per hook registration (EventType+YourURL).
```
Comment (maro-vm) [inline request-authentication.md]: This is not true, it's unique per webhook registration, which may cover multiple event types.

File: request-authentication.md [hunk 3]
```diff
@@ -6,13 +6,85 @@ pagination_next: null
...
+const secret = 'A0+AeKBRG2KRGvnNwJpQlb6IJFk48CKXCIcrLoHncVJKDILsQSxS6NWCccwWm6r6FhGKhiHTBsG2wo/xU6FY/A=='
+
+// The complete url of the webhook, including query parameters if applicable. I.e. https://yoursite.com/webhooks/vmp/v1
+const requestUrl = "https://webhook.site/e2cee29b-012e-4f1d-8ef4-e95fd74a7a63"
```
Comment (madfr-at-vipps) [inline request-authentication.md:36]: I think you should just leave this as a placeholder and indicate they should replace it instead. `https://webhook.site` is an external company that may or may not work or behave as we expect in the future, so we shouldn't couple our documentation to that. 

As an example, they have a restriction to number of incoming request per token (like the hardcoded one in the url), so if a few copy this then it sudden don't work for some.

File: request-authentication.md [hunk 4]
```diff
@@ -6,13 +6,85 @@ pagination_next: null
...
+
+To run: `node webhook.mjs`
+```js
+import crypto from "crypto";
+
+// The secret is found in the response body when you register a webhook. 
+// Note: The secret is unique per hook registration (EventType+YourURL).
+const secret = 'A0+AeKBRG2KRGvnNwJpQlb6IJFk48CKXCIcrLoHncVJKDILsQSxS6NWCccwWm6r6FhGKhiHTBsG2wo/xU6FY/A=='
+
+// The complete url of the webhook, including query parameters if applicable. I.e. https://yoursite.com/webhooks/vmp/v1
```
Comment (maro-vm) [inline request-authentication.md]: I think this comment can be removed, it should be pretty self explanatory in my opinion.
An url is an url, it's not the host name or something like that and it's further emphasized with the example.

File: request-authentication.md [hunk 5]
```diff
@@ -6,13 +6,85 @@ pagination_next: null
...
+const requestUrl = "https://webhook.site/e2cee29b-012e-4f1d-8ef4-e95fd74a7a63"
+// Part of the HTTP request header from Vipps MobilePay, found in the field 'x-ms-date'
+const dateTime = "Thu, 30 Mar 2023 08:38:32 GMT"
+
+// The signature/authorization sent by Vipps MobilePay for you to verify against. 
+// Part of the HTTP request header from Vipps MobilePay, found in the field 'authorization'
+const actualSignature = "HMAC-SHA256 SignedHeaders=x-ms-date;host;x-ms-content-sha256&Signature=agAiSyogQbDHpeucoNwYz+yAr5nJ+v+zasdkSbqzv+U="
+
+// HTTP Body
+const payloadJson = {
```
Comment (maro-vm) [inline request-authentication.md]: Variable name might be a little misleading, it's not necessarily json.
Suggestion, rename to httpBody (that way you can also remove the comment.

File: request-authentication.md [hunk 6]
```diff
@@ -6,13 +6,85 @@ pagination_next: null
...
+  "some-unique-content":"ee6e441b-cc4a-46f8-895d-a5af79bcc233/hello-world"
+}
+
+const hashedPayload =  crypto.createHash("sha256").update(JSON.stringify(payloadJson)).digest("base64")
+// returns 'lNlsp1XA03N34HrQsVzPgJKtC+r7l/RBF4V3JQUWMj4='
+
+// The complete url of the call to the webhook, including query parameters if applicable
```
Comment (maro-vm) [inline request-authentication.md]: Duplicate comment + see previous, I'd suggest removing it entirely, query params are part of an url.

File: request-authentication.md [hunk 7]
```diff
@@ -6,13 +6,85 @@ pagination_next: null
...
+const url = new URL(requestUrl)
+const pathAndQuery = url.pathname + url.search; 
+// returns '/e2cee29b-012e-4f1d-8ef4-e95fd74a7a63'
+const host = url.host 
+// returns 'webhook.site'
```
Comment (maro-vm) [inline request-authentication.md]: Suggestion, keep comments on the right side or above for blocks,
it's confusing to read if it's below too.

Also, I'm confused by 'returns', nothing returns here, it's 'evaluates too',
but the common format I know is just `const host = url.host // webhook.site`

File: request-authentication.md [hunk 8]
```diff
@@ -6,13 +6,85 @@ pagination_next: null
...
+
+// Example is using HTTP POST, hence starting with 'POST\n'
```
Comment (maro-vm) [inline request-authentication.md]: All webhook calls are POST, so this is a bit misleading to write 'example is using'.

File: request-authentication.md [hunk 9]
```diff
@@ -6,13 +6,85 @@ pagination_next: null
...
+const expectedSignedString = `POST\n${pathAndQuery}\n${dateTime};${host};${hashedPayload}`
+/* returns:
+POST
+/e2cee29b-012e-4f1d-8ef4-e95fd74a7a63
+Thu, 30 Mar 2023 08:38:32 GMT;webhook.site;lNlsp1XA03N34HrQsVzPgJKtC+r7l/RBF4V3JQUWMj4=
+*/
+
+// Run the HMAC algorithm on the expectedSignedString with the secret for the hook
```
Comment (maro-vm) [inline request-authentication.md]: HMAC algorithm -> HMAC SHA256 algorithm

File: request-authentication.md [hunk 10]
```diff
@@ -6,13 +6,85 @@ pagination_next: null
...
+const expectedSignature = crypto.createHmac("sha256", secret)
+  .update(expectedSignedString)
+  .digest("base64")
+// returns 'agAiSyogQbDHpeucoNwYz+yAr5nJ+v+zasdkSbqzv+U=' 
+// which matches the expected signature
```
Comment (maro-vm) [inline request-authentication.md]: Redundant comment, 'which matches the expected signature'

File: request-authentication.md [hunk 11]
```diff
@@ -6,13 +6,85 @@ pagination_next: null
...
+
+const expectedAuth = `HMAC-SHA256 SignedHeaders=x-ms-date;host;x-ms-content-sha256&Signature=${expectedSignature}`
+
+if(expectedAuth !== actualSignature) {
+  // Log the error.
+  // Fail, stop processing
+  throw new Error('Authorization fields do not match expected value!');
+}
+// Continue with business logic related to the webhook.
+console.log('Success! Business logic awaits!')
```
Comment (maro-vm) [inline request-authentication.md:75]: I think it's nice to formulate this as a working example, 
but I think you can take it a step further and start at 
the top with an actual node http request. 

E.g. define a method `isValidWebhookCallback(request)` that takes a request
and returns true or false depending on whether the request was valid.

Then create a dummy request object and use the method
```js
const isValidWebhookRequest = (request) => {
    ...
}

const incommingRequest = ...
assert.strictEqual(isValidWebhookRequest(incommingRequest), true)
```

**Review by maro-vm — COMMENTED** [review]
Neat to add a js example.
I gave you a few nitpicks.

In general I think there's way too many comments, especially given the preceding documentation. 

It also feels a bit off that we start with a module,
with hard coded values and then end we see something that looks like 
it could be production code 'Continue with business logic related to the webhook.'.

I'll defer to @cloveras on how the end result should look though. ^_^

File: request-authentication.md (hunk 11 above)
Comment (flinkgutt) [reply request-authentication.md:75]: Good idea, I initially created a remix.run app+ngrok where I tested this and some other stuff.
So I just pulled out the relevant code and tested that I could run it on the command line.

But this is a good idea "going forward" 😁

File: request-authentication.md (hunk 1 above)
Comment (flinkgutt) [reply request-authentication.md]: Removed in [e422710](https://github.com/vippsas/webhooks-api/pull/23/commits/e422710fb6dfc82aa8cbcbc2cb2ef26d4106aaf1)

File: request-authentication.md (hunk 2 above)
Comment (flinkgutt) [reply request-authentication.md]: Also fixed in [e422710](https://github.com/vippsas/webhooks-api/pull/23/commits/e422710fb6dfc82aa8cbcbc2cb2ef26d4106aaf1)

File: request-authentication.md (hunk 5 above)
Comment (flinkgutt) [reply request-authentication.md]: Changed to `httpBody` in [aee7d03](https://github.com/vippsas/webhooks-api/pull/23/commits/aee7d0391090df3f18f41fc243c6b40f8fa8dda4)

File: request-authentication.md (hunk 4 above)
Comment (flinkgutt) [reply request-authentication.md]: Removed comment in [8488443](https://github.com/vippsas/webhooks-api/pull/23/commits/84884437272f8c01c28dc7d0635a20ee7301a300)

File: request-authentication.md (hunk 8 above)
Comment (flinkgutt) [reply request-authentication.md]: Fixed, removed comment

File: request-authentication.md (hunk 9 above)
Comment (flinkgutt) [reply request-authentication.md]: Fixed

File: request-authentication.md (hunk 10 above)
Comment (flinkgutt) [reply request-authentication.md]: Fixed

File: request-authentication.md (hunk 7 above)
Comment (flinkgutt) [reply request-authentication.md]: Fixed for most cases

**Review by maro-vm — APPROVED** [review]
Looks good to me ^_^

File: request-authentication.md (hunk 11 above)
Comment (maro-vm) [reply request-authentication.md:75]:
```javascript
import crypto from "crypto";

function validate(url, dateTime, signature, secret, body) {

  const pathAndQuery = url.pathname + url.search; 
  // /e2cee29b-012e-4f1d-8ef4-e95fd74a7a63

  const host = url.host 
  // webhook.site

  const hashedPayload =  crypto
    .createHash("sha256")
    .update(JSON.stringify(httpBody))
    .digest("base64") 
    //'lNlsp1XA03N34HrQsVzPgJKtC+r7l/RBF4V3JQUWMj4=

  const expectedSignedString = `POST\n${pathAndQuery}\n${dateTime};${host};${hashedPayload}`
  // POST
  // /e2cee29b-012e-4f1d-8ef4-e95fd74a7a63
  // Thu, 30 Mar 2023 08:38:32 GMT;webhook.site;lNlsp1XA03N34HrQsVzPgJKtC+r7l/RBF4V3JQUWMj4=
  
  // Run the HMAC SHA256 algorithm on the expectedSignedString with the secret for the hook
  const expectedSignature = crypto
    .createHmac("sha256", secret)
    .update(expectedSignedString)
    .digest("base64")
    // agAiSyogQbDHpeucoNwYz+yAr5nJ+v+zasdkSbqzv+U=

  const validSignature = `HMAC-SHA256 SignedHeaders=x-ms-date;host;x-ms-content-sha256&Signature=${expectedSignature}`

  return validSignature !== actualSignature;
}

// The secret is found in the response body when you register a webhook.
// Note: The secret is unique per hook registration, which may cover multiple event types.
const secret = 'A0+AeKBRG2KRGvnNwJpQlb6IJFk48CKXCIcrLoHncVJKDILsQSxS6NWCccwWm6r6FhGKhiHTBsG2wo/xU6FY/A=='

const requestUrl = new URL("https://webhook.site/e2cee29b-012e-4f1d-8ef4-e95fd74a7a63")

// Part of the HTTP request header from Vipps MobilePay, found in the field 'x-ms-date'
const dateTime = "Thu, 30 Mar 2023 08:38:32 GMT"

// The signature/authorization sent by Vipps MobilePay for you to verify against. 
// Part of the HTTP request header from Vipps MobilePay, found in the field 'authorization'
const actualSignature = "HMAC-SHA256 SignedHeaders=x-ms-date;host;x-ms-content-sha256&Signature=agAiSyogQbDHpeucoNwYz+yAr5nJ+v+zasdkSbqzv+U="

const httpBody = { "some-unique-content":"ee6e441b-cc4a-46f8-895d-a5af79bcc233/hello-world" }

const isThisAValidWebhookRequest = validate(requestUrl, dateTime, actualSignature, secret, httpBody)
```

**Pull request merged**